File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/datasource/DataSourceSelectorImpl.scala
package net.gree.aurora.scala.domain.datasource

import net.gree.aurora.domain.datasource.{DataSource => JDataSource, JDBCDataSource => JJDBCDataSource, RedisDataSource => JRedisDataSource}
import net.gree.aurora.domain.datasource.{DataSourceSelector => JDataSourceSelector}
import org.sisioh.dddbase.utils.{Try => JTry}
import scala.util.{Failure, Success, Try}

private[domain]
class DataSourceSelectorImpl[T](underlying: JDataSourceSelector[T])
  extends DataSourceSelector[T] {

  private def toJava[T](hint: Option[T]):T = hint match {
    case Some(v) => v
    case None => null.asInstanceOf[T]
  }
  def selectDataSource(hint: Option[T]): Try[DataSource] = {
    val result = underlying.selectDataSource(toJava(hint))
    result match {
      case success: JTry.Success[_] =>
        Success(DataSource(success.get().asInstanceOf[JDataSource]))
      case failure: JTry.Failure[_] =>
        Failure(failure.getCause)
    }
  }

  def selectDataSourceAsJDBC(hint: Option[T]): Try[JDBCDataSource] = {
    val result = underlying.selectDataSourceAsJDBC(toJava(hint))
    result match {
      case success: JTry.Success[_] =>
        Success(JDBCDataSource(success.get().asInstanceOf[JJDBCDataSource]))
      case failure: JTry.Failure[_] =>
        Failure(failure.getCause)
    }
  }

  def selectDataSourceAsRedis(hint: Option[T]): Try[RedisDataSource] = {
    val result = underlying.selectDataSourceAsRedis(toJava(hint))
    result match {
      case success: JTry.Success[_] =>
        Success(RedisDataSource(success.get().asInstanceOf[JRedisDataSource]))
      case failure: JTry.Failure[_] =>
        Failure(failure.getCause)
    }
  }

}

File: aurora-core/src/main/java/net/gree/aurora/domain/datasource/DataSourceSelectorImpl.java
package net.gree.aurora.domain.datasource;

import org.apache.commons.lang.Validate;
import org.sisioh.dddbase.utils.Function1;
import org.sisioh.dddbase.utils.Try;

import java.util.ArrayList;
import java.util.List;

final class DataSourceSelectorImpl<T> implements DataSourceSelector<T> {

    private final List<DataSourceId> targetIds;
    private final DataSourceRepository dataSourceRepository;
    private final DataSourceIdResolver<T> dataSourceResolver;

    DataSourceSelectorImpl(
            DataSourceRepository dataSourceRepository,
            DataSourceIdResolver<T> dataSourceResolver
    ) {
        Validate.notNull(dataSourceRepository);
        Validate.notNull(dataSourceResolver);
        this.dataSourceRepository = dataSourceRepository;
        targetIds = new ArrayList<>();
        for (DataSource ds : dataSourceRepository.toList()) {
            targetIds.add(ds.getIdentity());
        }
        this.dataSourceResolver = dataSourceResolver;

    }

    DataSourceSelectorImpl(
            DataSourceRepository dataSourceRepository,
            List<DataSourceId> targetIds,
            DataSourceIdResolver<T> dataSourceResolver
    ) {
        Validate.notNull(dataSourceRepository);
        Validate.notNull(targetIds);
        Validate.notNull(dataSourceResolver);
        this.dataSourceRepository = dataSourceRepository;
        this.targetIds = targetIds;
        this.dataSourceResolver = dataSourceResolver;
    }

    @Override
    public <A extends DataSource> Try<A> selectDataSourceAs(final Class<A> clazz, T hint) {
        DataSourceId dataSourceId = dataSourceResolver.apply(hint);
        if (!targetIds.contains(dataSourceId)) {
            throw new IllegalArgumentException();
        }
        return dataSourceRepository.resolve(dataSourceId).map(new Function1<DataSource, A>() {
            @Override
            public A apply(DataSource value) {
                return clazz.cast(value);
            }
        });
    }

    @Override
    public Try<DataSource> selectDataSource(T hint) {
        return selectDataSourceAs(DataSource.class, hint);
    }

    @Override
    public Try<JDBCDataSource> selectDataSourceAsJDBC(T hint) {
        return selectDataSourceAs(JDBCDataSource.class, hint);
    }

    @Override
    public Try<RedisDataSource> selectDataSourceAsRedis(T hint) {
        return selectDataSourceAs(RedisDataSource.class, hint);
    }

}
